go/libkb/chain_link_parse_2.go

```diff
@@ -0,0 +1,623 @@
+			}
+		}
+	} else {
+		highSkipHash, err = LinkIDFromHex(highSkipHashStr)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	highSkip := NewHighSkip(keybase1.Seqno(highSkipSeqnoInt), highSkipHash)
+	return &highSkip, nil
+}
+
+func newSigInfo(kid keybase1.KID, payload []byte, sig kbcrypto.NaclSignature) *kbcrypto.NaclSigInfo {
+	return &kbcrypto.NaclSigInfo{
+		Kid:      kid.ToBinaryKID(),
+		Payload:  payload,
+		Sig:      sig,
+		SigType:  kbcrypto.SigKbEddsa,
+		HashType: kbcrypto.HashPGPSha512,
+		Detached: true,
+	}
+}
+
+func decodeSig1Imploded(s string) (*kbcrypto.NaclSignature, error) {
+	raw, err := base64.StdEncoding.DecodeString(s)
+	if err != nil {
+		return nil, err
+	}
+	var ret kbcrypto.NaclSignature
+	copy(ret[:], raw)
+	return &ret, nil
+}
+
+func importLinkFromServerV1NaCl(m MetaContext, packed []byte) (*importRes, error) {
+	var sigBody []byte
+	var ret importRes
+
+	sig1ImplodedRaw := jsonGetString(packed, "si1")
+	if sig1ImplodedRaw == "" {
+		return nil, ChainLinkError{"no si1 field as expected"}
+	}
+
+	sig1Imploded, err := decodeSig1Imploded(sig1ImplodedRaw)
+	if err != nil {
+		return nil, err
+	}
+
+	payloadJSON, err := getPayloadJSONFromServerLink(packed)
+	if err != nil {
+		return nil, err
+	}
+	version, err := payloadJSON.Version()
+	if err != nil {
+		return nil, err
+	}
+	if version != KeybaseSignatureV1 {
+		return nil, ChainLinkError{"inner chainlink showed wrong version, while expecting 1"}
+	}
+	ret.linkID = payloadJSON.Hash()
+	err = payloadJSON.AssertJSON(ret.linkID)
+	if err != nil {
+		return nil, err
+	}
+	ret.kid, err = payloadJSON.KID()
+	if err != nil {
+		return nil, err
+	}
+	sigInfo := newSigInfo(ret.kid, payloadJSON.Bytes(), *sig1Imploded)
+	clientName, clientVersion := payloadJSON.ClientNameAndVersion()
+	sigBody, ret.sigID, err = sigid.ComputeSigBodyAndID(sigInfo, clientName, clientVersion)
+	if err != nil {
+		return nil, err
+	}
+	ret.sig = base64.StdEncoding.EncodeToString(sigBody)
+	ret.payload = payloadJSON.Bytes()
+	return &ret, nil
+}
+
+type sig2Imploded struct {
+	_struct   bool `codec:",toarray"` //nolint
+	Sig       kbcrypto.NaclSignature
+	OuterLink OuterLinkV2
+	NumFields int
+}
+
+type sigChainPayloadJSON struct {
+	s string
+	b []byte
+}
+
+func newSigChainPayloadJSON(s string) *sigChainPayloadJSON {
+	return &sigChainPayloadJSON{s: s, b: []byte(s)}
+}
+
+func newSigChainPayloadJSONFromBytes(b []byte) *sigChainPayloadJSON {
+	return &sigChainPayloadJSON{s: string(b), b: b}
+}
+
+func (s *sigChainPayloadJSON) String() string {
+	return s.s
+}
+
+func (s *sigChainPayloadJSON) Bytes() []byte {
+	return s.b
+}
+
+func (s *sigChainPayloadJSON) Hash() LinkID {
+	return ComputeLinkID(s.b)
+}
+
+func getPayloadJSONFromServerLink(packed []byte) (*sigChainPayloadJSON, error) {
+	data, _, _, err := jsonparserw.Get(packed, "payload_json")
+	if err != nil {
+		return nil, err
+	}
+	sdata, err := strconv.Unquote(`"` + string(data) + `"`)
```

Comment:
as long as we're imploding stuff, does it make sense to deserialize payload_json on the server to save on all this escaping? or would it be a perf hit?

Reply:
hmmm, then we'll have to worry about the canonicalization of the JSON, but it should be canonicalized....